File: 2001~2500/2265. Count Nodes Equal to Average of Subtree/main.py
"""
Source: https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/
Date: 2022/12/2
Skill: 
Runtime: 91 ms, faster than 59.12%
Memory Usage: 15.2 MB, less than 35.70%
Time complexity: 
Space complexity: 
Constraints: 
    The number of nodes in the tree is in the range [1, 1000].
    0 <= Node.val <= 1000
"""

from typing import List, Optional
from collections import Counter

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        res = 0
        
        def post_order_trace(root):
            nonlocal res
            if not root:
                return 0, 0
            
            left_sum, left_count = post_order_trace(root.left)
            right_sum, right_count = post_order_trace(root.right)
            if (left_sum+right_sum+root.val)//(left_count+right_count+1) == root.val:
                res += 1
            return left_sum+right_sum+root.val, left_count+right_count+1
                
        post_order_trace(root)
        return res

File: 2001~2500/2265. Count Nodes Equal to Average of Subtree/main.cpp
/**
 * Source: t.ly/AW6EQ
 * Date: 2023/11/2
 * Skill:
 * Runtime: 3 ms, faster than 92.80% of C++ online submissions
 * Memory Usage: 12.49 MB, less than 21.30% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


typedef pair<int, int> pairs;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int averageOfSubtree(TreeNode* root) {
        int res = 0;
        function<pairs(TreeNode *)> dfs;
        dfs = [&](TreeNode *node) {
            if (node == nullptr) return make_pair(0, 0);
            int sum = node->val, num = 1;
            auto[s1, n1] = dfs(node->left);
            auto[s2, n2] = dfs(node->right);
            sum += s1 + s2;
            num += n1 + n2;
            if (node->val  == sum / num) res++;
            return make_pair(sum, num);
        };
        dfs(root);
        return res;
    }
};


int main() {
    Solution s;
}